### ISSUE TITLE
Interface to `<: Supervised` MLJ models

### ISSUE THREAD

**Issue by pat-alt** [JuliaTrustworthyAI/CounterfactualExplanations.jl#69]
*This is an issue reserved for the [TU Delft Student Software Project '23](https://github.com/orgs/JuliaTrustworthyAI/projects/4)*

[MLJ](https://alan-turing-institute.github.io/MLJ.jl/dev/) is a popular machine learning framework for Julia. It ships with a large suite of common machine learning models, both supervised and unsupervised. It seems natural to interface this package to MLJ, although currently differentiability is a major challenge: to be able to use any of our counterfactual generators to explain MLJ models, those models need to be differentiable with respect to features. Still, this is worth exploring.

I propose the following steps:

1. Implement basic interface to MLJ (essentially have an `AbstractFittedModel` for `MLJ.Supervised`)
2. From the MLJ [model list](https://alan-turing-institute.github.io/MLJ.jl/dev/list_of_supported_models/), identify which ML models fulfil the differentiability criterium. Note that some models, like decision trees, may be differentiable after [probability calibration](https://scikit-learn.org/stable/modules/calibration.html). See below for a potential starting point. Start by focusing on pure Julia models, before dealing with non-native models (like sklearn).
3. Ideally, I think we would like a single `MLJModel<:AbstractFittedModel` class that can handle all (compatible) supervised MLJ models. To this end, we will need a mechanism to differentiate between compatible and incompatible models.
4. Thoroughly test and document your contributions. 

This is a challenging task and it is not critical that you succeed at everything. But we would like to aim for the following minimum achievements:

- [x] Add the basic interface (point 1)
- [x] Document your process and findings regarding point 2
- [ ] If a complete interface turns out to be too challenging, work on a proof-of-concept at least for one particular MLJ model, ideally Evotrees (points 3 and 4)

### Previous attempts

I have tried this in the past, which might or might not be a good starting point:

1. At this point all of the counterfactual generators need gradient-access and currently leverage `Zygote.jl` for auto-diff. Not sure if all MLJ models can just be "auto-diffed" in that sense, but some early experiments with `EvoTrees` has shown that in principal gradient-based counterfactual generators should be applicable (see [here](https://github.com/pat-alt/CounterfactualExplanations.jl/blob/037b3a4f0d5adba79c1137cb25cca6459cca46c0/src/models/differentiable/julia.jl#L100)). 
5. That being said, `Zygote.jl` didn't work in this case and I had to rely on `ForwardDiff` (see [here](https://github.com/JuliaTrustworthyAI/CounterfactualExplanations.jl/blob/037b3a4f0d5adba79c1137cb25cca6459cca46c0/src/generators/gradient_based/functions.jl#L23)). The problem with trees is that the counterfactual loss function is not smooth and hence taking gradients just resulted in gradients with all elements equal to zero (at least I think the non-smoothness was the issue here). Would still be preferable to use Zygote if possible.
6. (Non-)Differentiability of models may be a more general issue.

**Comment by pat-alt** [comment]
`MLJFlux` is probably the most obvious place to start for this (see related discussion [here](https://github.com/FluxML/MLJFlux.jl/issues/220))

**Comment by pat-alt** [comment]
This is in principle now implemented (#450), but by default MLJ models are assumed to be non-differentiable (the `MLJBase.predict` call and other functions don't play nicely with Zygote)
